client/service/src/lib.rs

```diff
@@ -196,178 +107,80 @@ pub trait AbstractService: Future<Output = Result<(), Error>> + Send + Unpin + S
+/// For each element, every time the `Interval` fires we push an element on the sender.
+pub struct NetworkStatusSinks<Block: BlockT>(
+	Arc<Mutex<status_sinks::StatusSinks<(NetworkStatus<Block>, NetworkState)>>>,
+);
+
+impl<Block: BlockT> NetworkStatusSinks<Block> {
+	fn new(
+		sinks: Arc<Mutex<status_sinks::StatusSinks<(NetworkStatus<Block>, NetworkState)>>>
+	) -> Self {
+		Self(sinks)
 	}
 
-	fn network_status(&self, interval: Duration) -> TracingUnboundedReceiver<(NetworkStatus<Self::Block>, NetworkState)> {
+	/// Returns a receiver that periodically receives a status of the network.
+	pub fn network_status(&self, interval: Duration)
+		-> TracingUnboundedReceiver<(NetworkStatus<Block>, NetworkState)> {
 		let (sink, stream) = tracing_unbounded("mpsc_network_status");
-		self.network_status_sinks.lock().push(interval, sink);
+		self.0.lock().push(interval, sink);
 		stream
 	}
-
-	fn transaction_pool(&self) -> Arc<Self::TransactionPool> {
-		self.transaction_pool.clone()
-	}
-
-	fn on_exit(&self) -> exit_future::Exit {
-		self.task_manager.on_exit()
-	}
-
-	fn prometheus_registry(&self) -> Option<prometheus_endpoint::Registry> {
-		self.prometheus_registry.clone()
-	}
-
-	fn base_path(&self) -> Option<Arc<BasePath>> {
-		self._base_path.clone()
-	}
 }
 
-impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Future for
-	Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc>
-{
-	type Output = Result<(), Error>;
-
-	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
-		let this = Pin::into_inner(self);
+/// Sinks to propagate telemetry connection established events.
+pub struct TelemetryOnConnectSinks(pub Arc<Mutex<Vec<TracingUnboundedSender<()>>>>);
 
-		match Pin::new(&mut this.essential_failed_rx).poll_next(cx) {
-			Poll::Pending => {},
-			Poll::Ready(_) => {
-				// Ready(None) should not be possible since we hold a live
-				// sender.
-				return Poll::Ready(Err(Error::Other("Essential task failed.".into())));
-			}
-		}
-
-		// The service future never ends.
-		Poll::Pending
+impl TelemetryOnConnectSinks {
+	/// Get event stream for telemetry connection established events.
+	pub fn on_connect_stream(&self) -> TracingUnboundedReceiver<()> {
+		let (sink, stream) =tracing_unbounded("mpsc_telemetry_on_connect");
+		self.0.lock().push(sink);
+		stream
 	}
 }
 
-impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Spawn for
-	Service<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc>
-{
-	fn spawn_obj(
-		&self,
-		future: FutureObj<'static, ()>
-	) -> Result<(), SpawnError> {
-		self.task_manager.spawn_handle().spawn("unnamed", future);
-		Ok(())
-	}
+/// The individual components of the chain, built by the service builder. You are encouraged to
+/// deconstruct this into its fields.
+pub struct ServiceComponents<
+	TBl: BlockT, TBackend: sc_client_api::backend::Backend<TBl>, TExec, TRtApi, TSc, TExPool,
+> {
+	/// A blockchain client.
+	pub client: Arc<Client<TBackend, TExec, TBl, TRtApi>>,
```

Comment:
ok so didn't want to say anything till I confirmed with @gnunicorn, but the `Client` struct was removed from our public api in #5502, so it doesn't really make sense to re-introduce it here.

so what we could do instead is make the client field simply be a generic `Arc<Client>` then add the proper `where` bounds to `ServiceComponents`'s definition allowing it to behave _like_ the `Client` struct but with the advantage that people don't depend on any substrate-internal struct methods.

cc-ing @gnunicorn in case he has anything to add.

Meanwhile I'm currently using this new interface for [paritytech/substrate-test-runner](https://github.com/paritytech/substrate-test-runner) and it is simply a joy to use 😋, can't wait to get this PR merged. :clap:

Reply:
When you say "behave like the Client" you mean:
```
		HeaderMetadata<Self::Block, Error = sp_blockchain::Error> + UsageProvider<Self::Block>
		+ BlockchainEvents<Self::Block> + HeaderBackend<Self::Block> + Send + Sync;
```
?

I wonder if that would help the situation I have here: #6521 I need to import blocks but I can't find a proper way to do it :thinking:

Reply:
So after briefly looking at the error in your PR desc, it looks to me like `PolkadotClient` is missing the trait bound for `BlockImport`, :sweat_smile: 


```rust
pub trait PolkadotClient<Block, Backend, Runtime>:
	BlockchainEvents<Block> + Sized + Send + Sync
	+ ProvideRuntimeApi<Block, Api = Runtime::RuntimeApi>
	+ HeaderBackend<Block>
	+ CallApiAt<
		Block,
		Error = sp_blockchain::Error,
		StateBackend = Backend ::State
	>
	where
		Block: BlockT,
		Backend: BackendT<Block>,
		Runtime: ConstructRuntimeApi<Block, Self>
{}
```

adding it should give you the `import` method on `ClientBlockImportExt`

Reply:
on second thought, you might need to do some `Deref` juju,

```rust
(&*alice.client).import((), ())
```

Reply:
haha :joy: Good one! .... nah :sob:  thanks for having a look anyway. I think I have an alternative solution so I will close that PR for now, wait for Ashley to merge this one, and see later how I can improve.

Reply:
subtrate code example here:

https://github.com/paritytech/substrate/blob/a273b48a0f32e7f7a670d3698453e3249521865b/client/finality-grandpa/src/import.rs#L59

used as:

https://github.com/paritytech/substrate/blob/a273b48a0f32e7f7a670d3698453e3249521865b/client/finality-grandpa/src/import.rs#L440

hopefully I've been helpful :sweat_smile:

Reply:
Just tested now...
```
error[E0596]: cannot borrow data in a `&` reference as mutable
  --> polkadot-test-service/tests/build-blocks.rs:52:3
   |
52 |         (&*alice.client).import(block).unwrap(); // TODO
   |         ^^^^^^^^^^^^^^^^ cannot borrow as mutable
```
But it's okay don't worry for now. I will send a transaction via RPC and find a better solution later.

Reply:
@seunlanlege Done that now.